### PR TITLE
Make HTTP/2 optional via 'pip install httpx[http2]'

### LINKED ISSUE #496: Require `pip install httpx[http2]` for HTTP/2 support
I wonder if we should consider having HTTP/2 be an optional requirement, rather than always-on.

I've been working towards minimal installs in other projects eg. Starlette (no hard dependencies) and Uvicorn - we're planning to move towards `pip install uvicorn` for a basse install of just `click` and `h11`, and `pip install uvicorn[default]` to pull in websockets plus platform specfic stuff like uvloop and httptools. And generally with the other projects in the encode async space too.

We *could* consider having minimal and default install options with httpx too... eg:

* `pip install httpx` - Just http/1.1 support.
* `pip install httpx[defaults]` - Also http/2. Potentially http/3 and/or websocket support at a future date. Possible that hsts isn't strictly a hard requirement either.

Any thoughts?

### PR DESCRIPTION
**Pull request by florimondmanca**
Fixes #496 

This PR makes HTTP/2 support entirely optional, and requires installing with `pip install httpx[http2]` to use HTTP/2, in particular to pass `http2=True` to a `Client`.

Note that our development setup always installs HTTP/2 dependencies. If we go and `import h2` again in the future in places where it shouldn't be installed, we won't catch the issue locally nor on CI. I think a good option might be to add a "Test with a base install" job to CI, but we'll need to shuffle things around in `requirements.txt` and our install scripts to make that happen. Happy to treat it as a follow-up. :-)

### CODE REVIEW

**Comment by florimondmanca** [comment]
This is *technically* a breaking change (although users may not catch it immediately because they'll already have `h2` installed locally, so they'll only see `MissingDependencies` on CI or a fresh install), and I believe there's no way of being backwards-compatible here. The error message is clear enough that I might be OK with releasing this as part of 0.10, but if we stick to discussions in #587 then this will need to wait for 1.0?

File: httpx/__init__.py [hunk 1]
```diff
@@ -58,6 +59,7 @@
     "DecodingError",
     "HTTPError",
     "InvalidURL",
+    "MissingDependencies",
```
Comment (tomchristie) [inline httpx/__init__.py]: I don't think we should introduce exception classes for this sort of case. A `RuntimeError` or something similar would be suitable here.

Comment (florimondmanca) [reply httpx/__init__.py]: Fair enough :-) Dropped this exception class in the latest commit.

**Comment by florimondmanca** [comment]
The build's broken due to a crazy `ImportError` caused by `regex`: https://travis-ci.org/encode/httpx/jobs/625866049. Not able to reproduce locally. 😕 

```console
$ scripts/check
+black --check --diff --target-version=py36 httpx tests
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/bin/black", line 6, in <module>
    from black import patched_main
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/black.py", line 15, in <module>
    import regex as re
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/regex.py", line 402, in <module>
    import _regex_core
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/_regex_core.py", line 21, in <module>
    import regex._regex as _regex
ModuleNotFoundError: No module named 'regex._regex'; 'regex' is not a package
```

The `regex` folks are experimenting with wheels for Linux (see https://bitbucket.org/mrabarnett/mrab-regex/issues/343/wheel-for-linux) so let me try tweaking our `install` script for now…

File: httpx/extras.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+"""
```
Comment (tomchristie) [inline httpx/extras.py]: I'm wondering about `optionals.py`?

Comment (florimondmanca) [reply httpx/extras.py]: Sure, I like that name too. :-)

Comment (florimondmanca) [reply httpx/extras.py]: Sure, I can roll with that. :-)

**Comment by tomchristie** [comment]
I'm holding back on this slightly, because I feel like we ought to weigh up if we're being overly heavyhanded here. "We've got this ace new HTTP/2 feature. Btw, it's only much use on async. And it's not installed by default. Any you need to turn it on explicitly even once it is installed."

Not totally sure either way, but feeling like *maybe* we don't want to jump on it just yet, for 0.10. ?

**Comment by florimondmanca** [comment]
> We've got this ace new HTTP/2 feature. Btw, it's only much use on async. And it's not installed by default. Any you need to turn it on explicitly even once it is installed.

Put like that, it sure starts to sound like we don't even *want* users to use HTTP/2? 🤣

Maybe the `http2` flag is a strong enough barrier-to-usage already.

An alternative experience we might want to offer is to automatically switch `http2` on if we detect HTTP/2 dependencies are there — i.e. have `http2="auto"` be the default? That way users only have to `pip install httpx[http2]`, and they'll have HTTP/2 enabled when possible, right away.

**Comment by tomchristie** [comment]
> An alternative experience we might want to offer is to automatically switch http2 on if we detect HTTP/2 dependencies are there

That's probably a no. It's too painful that HTTP behavior might switch depending on *other* packages being installed. (Eg. A user installs hypercorn or uvicorn and suddenly their `httpx` behavior changes.)

**Comment by tomchristie** [comment]
Closing as per #496

**Pull request closed without merging**